model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/locking/HotRodLocksUtils.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2022 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.models.map.storage.hotRod.locking;
+
+import org.infinispan.client.hotrod.Flag;
+import org.infinispan.client.hotrod.RemoteCache;
+import org.keycloak.models.locking.LockAcquiringTimeoutException;
+
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+
+public class HotRodLocksUtils {
+
+    public static final String HOT_ROD_LOCKS_CACHE_NAME = "locks";
+
+    /**
+     * Repeatedly attempts to put an entry with the key {@code lockName}
+     * to the {@code cache}. Succeeds only if there is no entry with the same
+     * key already.
+     * <p />
+     * Execution of this method is time bounded, if this method does not
+     * succeed within {@code timeoutMilliseconds} it gives up and returns
+     * false.
+     * <p />
+     * There is a pause after each unsuccessful attempt equal to
+     * {@code repeatInterval} milliseconds
+     *
+     * @param cache Cache that will be used for putting the value
+     * @param lockName Name of the entry
+     * @param timeoutMilliseconds time boundaries in milliseconds
+     * @param repeatInterval Number of milliseconds to wait after each
+     *                       unsuccessful attempt
+     *
+     * @return true when the key {@code lockName} was put into the {@code cache}
+     *         within time boundaries and previous value was {@code null},
+     *         false otherwise
+     */
+    public static boolean repeatPutIfAbsent(RemoteCache<String, String> cache, String lockName, long timeoutMilliseconds, int repeatInterval) throws LockAcquiringTimeoutException {
+        String currentOwner;
+
+        while((currentOwner = cache.withFlags(Flag.FORCE_RETURN_VALUE)
+                .putIfAbsent(lockName, getKeycloakInstanceIdentifier(), 900000, java.util.concurrent.TimeUnit.DAYS)) != null
+                && timeoutMilliseconds > 0) {
+            try {
+                Thread.sleep(repeatInterval);
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+            timeoutMilliseconds -= repeatInterval;
+        }
+
+        return currentOwner == null;
+    }
+
+    private static String getKeycloakInstanceIdentifier() {
+        try {
+            return InetAddress.getLocalHost().getHostName();
```

Comment:
Do I understand this correctly, that in model tests where each node name is the same (in my case "localhost.localdomain") one node might accidentally unlock other's node lock if it knows the lock name and has access to the lock's reference. Probably not big deal anyway.

Reply:
Yes, you understand it correctly. I think this is what we want based on some discussions we had, but would be good to have @hmlnarik 's opinion on this